src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs

```diff
@@ -2742,74 +2740,145 @@ private static IEnumerable<string> CompleteFileVerbs(string filePath, string wor
         /// JobObjectHandle is a reference to the job object used to track
         /// the child processes created by the main process hosted by the Start-Process cmdlet.
         /// </summary>
-        private readonly Microsoft.PowerShell.Commands.SafeJobHandle _jobObjectHandle;
+        private nint _jobObjectHandle;
+
+        /// <summary>
+        /// The completion port handle that is used to monitor job events.
+        /// </summary>
+        private nint _completionPortHandle;
 
         /// <summary>
         /// ProcessCollection constructor.
         /// </summary>
         internal ProcessCollection()
-        {
-            IntPtr jobObjectHandleIntPtr = NativeMethods.CreateJobObject(IntPtr.Zero, null);
-            _jobObjectHandle = new SafeJobHandle(jobObjectHandleIntPtr);
-        }
+        {}
 
         /// <summary>
         /// Start API assigns the process to the JobObject and starts monitoring
         /// the child processes hosted by the process created by Start-Process cmdlet.
         /// </summary>
+        /// <returns>Whether the job and assignment worked or not.</returns>
         internal bool AssignProcessToJobObject(SafeProcessHandle process)
         {
-            // Add the process to the job object
-            bool result = Interop.Windows.AssignProcessToJobObject(
-                _jobObjectHandle.DangerousGetHandle(),
+            if (!InitializeJob())
+            {
+                return false;
+            }
+
+            // // Add the process to the job object
+            return Interop.Windows.AssignProcessToJobObject(
+                this._jobObjectHandle,
                 process.DangerousGetHandle());
-            return result;
         }
 
         /// <summary>
-        /// Checks to see if the JobObject is empty (has no assigned processes).
-        /// If job is empty the auto reset event supplied as input would be set.
+        /// WaitOne blocks the current thread until the job receives a
+        /// completion notification.
         /// </summary>
-        internal void CheckJobStatus(object stateInfo)
+        /// <param name="waitHandleToUse">
+        /// WaitHandle to use for waiting on the job object.
+        /// </param>
+        internal void WaitOne(ManualResetEvent waitHandleToUse)
         {
-            ManualResetEvent emptyJobAutoEvent = (ManualResetEvent)stateInfo;
-            int dwSize = 0;
-            const int JOB_OBJECT_BASIC_PROCESS_ID_LIST = 3;
-            JOBOBJECT_BASIC_PROCESS_ID_LIST JobList = new();
+            if (this._completionPortHandle == nint.Zero)
+            {
+                return;
+            }
+
+            const int INFINITE = -1;
+            int completionCode = 0;
+            do
+            {
+                Interop.Windows.GetQueuedCompletionStatus(
+                    this._completionPortHandle,
+                    out completionCode,
+                    out _,
+                    out _,
+                    INFINITE);
```

Comment:
> INFINITE

Will does Ctrl-C still kill the cmdlet/pipeline?

Reply:
That’s a good point, I never hooked that up.

Reply:
This SO post contains a reasonable-looking summary of possible ways to wake up the thread: https://stackoverflow.com/questions/32921513/how-to-force-getqueuedcompletionstatus-to-return-immediately

Reply:
Given that we only have a single thread on the completion port, I think the cleanest option is to use `PostQueuedCompletionStatus` and post a custom completion to the IOCP.